ocs_ci/utility/prometheus.py

```diff
@@ -172,6 +175,68 @@ def get(self, resource, payload=None):
         )
         return response
 
+    def query_range(self, query, start, end, step, timeout=None, validate=True):
+        """
+        Perform Prometheus `range query`_. This is a simple wrapper over
+        ``get()`` method with plumbing code for range queries, additional
+        validation and logging.
+
+        Args:
+            query (str): Prometheus expression query string.
+            start (str): start timestamp (rfc3339 or unix timestamp)
+            end (str): end timestamp (rfc3339 or unix timestamp)
+            step (float): Query resolution step width as float number of
+                seconds.
+            timeout (str): Evaluation timeout in duration format. Optional.
+            validate (bool): Perform basic validation on the response.
+                Optional, ``True`` is the default. Use ``False`` when you
+                expect query to fail eg. during negative testing.
+
+        Returns:
+            list: result of the query
+
+        .. _`range query`: https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries
+        """
+        query_payload = {
+            'query': query,
+            'start': start,
+            'end': end,
+            'step': step,
+            }
+        if timeout is not None:
+            query_payload['timeout'] = timeout
+        # Human readable summary of the query (details are logged by get
+        # method itself with debug level).
+        logger.info((
+            f"Performing prometheus range query '{query}' "
+            f"over a time range ({start}, {end})"))
+        resp = self.get('query_range', payload=query_payload)
+        content = yaml.safe_load(resp.content)
+        if validate:
+            # If this fails, Prometheus instance is so broken that test can't
+            # be performed.
+            assert content["status"] == "success"
+            # For a range query, we should always get a matrix result type, as
+            # noted in Prometheus documentation, see:
+            # https://prometheus.io/docs/prometheus/latest/querying/api/#range-vectors
+            assert content["data"]["resultType"] == "matrix"
+            # All metric sample series has the same size.
+            sizes = []
+            for metric in content["data"]["result"]:
+                sizes.append(len(metric["values"]))
+            assert all(size == sizes[0] for size in sizes)
```

Comment:
Worth adding an assertion message

Reply:
Also, could you please elaborate on why all sizes list items values need to be similar to each other?

Reply:
Isn't this sample rate?

Reply:
>  Isn't this sample rate?

No, because *sample rate* is usually understood as number of samples per some standard unit, such as second or minute. But here, this is just number of samples for the whole lifetime of a workload. If I normalized it to per minute or second, it could be called a sample rate. I'm not doing this because in general, I'm not interested in such value, but rather that I get the same number of sample for each sample series.

Reply:
> Also, could you please elaborate on why all sizes list items values need to be similar to each other?

When I ask Prometheus for values based on a query, most of the time I will get multiple sample series. Eg. for `ceph_osd_stat_bytes_used` I get a series of values for each osd. Unless I do something special with the cluster (eg. killing a node, expanding, ...) different number of values suggests that something unexpected happened. If I'm doing exactly that on purpose, I can disable this check by setting `validate=False` when calling the method.

Reply:
It's possible that this check turns out to be too restrictive, but in such case I can tweak it as necessary based on further experience.

Reply:
Message for assert statement included.